src/nvmet.c

```diff
@@ -833,209 +909,230 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
-				if (strlen(json_message) > 0) {
-					strcat(json_message, " ");
+			for (n = 0; n < port_number; n++) {
+				sprintf(path, "%s/%s", SYS_NVMET_PORTS, list[n]->d_name);
+				if (access(path, F_OK) == SUCCESS) {
+					sprintf(path, "%s/%s/subsystems/%s", SYS_NVMET_PORTS, list[n]->d_name, nqn);
+					if (access(path, F_OK) == SUCCESS) {
+						if (unlink(path) != SUCCESS) {
+							msg = "Error. Unable to remove NQN %s. %s: rmdir: %s";
+							print_error(msg, return_message, path, __func__, strerror(errno));
+							list = clean_scandir(list, port_number);
+							return rc;
+						}
+					}
 				}
-				strcat(json_message, json_message_temp);
 			}
-		} else {
-			/* The NVMe Target interface port number that was defined does not exist on the system */
-			msg = "%s: Port %d and / or export does not exist.";
-			print_error(msg, return_message, __func__, port);
-			return rc;
+			list = clean_scandir(list, port_number);
 		}
-	}
-
-	/*
-	 * If there are no more allowed_hosts and the target is not being exported by the port, remove
-	 * the target export from the entire subsystem. NOTE - we are not checking other ports (yet).
-	 * This check will come in a future update.
-	 */
-	if ((port != INVALID_VALUE) && (hostno == 0)) {
-		if (access(path, F_OK) == SUCCESS) {
-			msg = "Error. Target is still exported through the port: %s.";
-			print_error(msg, return_message, __func__, path);
+		sprintf(path, "%s/%s/attr_allow_any_host", SYS_NVMET_TGT, nqn);
+		if ((fp = fopen(path, "w")) == NULL){
+			msg = "Error. Unable to open %s. %s: fopen: %s";
+			print_error(msg, return_message, path, __func__, strerror(errno));
 			return INVALID_VALUE;
 		}
+		fprintf(fp, "0");
+		fclose(fp);
+
 		/* Disable volume */
-		sprintf(path, "%s/%s%s-%s/namespaces/1/enable", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+		sprintf(path, "%s/%s/namespaces/1/enable", SYS_NVMET_TGT, nqn);
 		if ((fp = fopen(path, "w")) == NULL){
 			msg = "Error. Unable to open namespace enable file %s. %s: fopen: %s";
 			print_error(msg, return_message, path, __func__, strerror(errno));
-			return INVALID_VALUE;
+			return rc;
 		}
 		fprintf(fp, "0");
 		fclose(fp);
 
 		/* Remove namespace */
-		sprintf(path, "%s/%s%s-%s/namespaces/1", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+		sprintf(path, "%s/%s/namespaces/1", SYS_NVMET_TGT, nqn);
 		if (access(path, F_OK) == SUCCESS) {
-			if ((rc = rmdir(path)) != SUCCESS) {
+			if (rmdir(path) != SUCCESS) {
 				msg = "Error. Unable to remove namespace %s. %s: rmdir: %s";
 				print_error(msg, return_message, path, __func__, strerror(errno));
 				return rc;
 			}
 		}
 
 		/* Remove NQN from NVMeT Subsystem */
-		sprintf(path, "%s/%s%s-%s", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+		sprintf(path, "%s/%s", SYS_NVMET_TGT, nqn);
 		if (access(path, F_OK) == SUCCESS) {
-			if ((rc = rmdir(path)) != SUCCESS) {
+			if (rmdir(path) != SUCCESS) {
+				sprintf(path, "%s/%s/namespaces/1", SYS_NVMET_TGT, nqn);
```

Comment:
@matteotenca This part confuses me. Why are we retrying to remove the namespace when it was already done above?

Reply:
This is a small hack: if the `rmdir` of `/sys/kernel/config/nvmet/<NQN>` fails for any reason, we try to re-create the namespace. If I remember well, I found out a situation where the `rmdir` of the NQN failed for good reasons, but since the namespace was gone, the tools could not handle the situation anymore. Putting back the namespace directory allowed to be still able to see the NQN.